flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -169,35 +170,103 @@ public Dispatcher(
             DispatcherBootstrapFactory dispatcherBootstrapFactory,
             DispatcherServices dispatcherServices)
             throws Exception {
+        this(
+                rpcService,
+                fencingToken,
+                recoveredJobs,
+                recoveredDirtyJobs,
+                dispatcherBootstrapFactory,
+                dispatcherServices,
+                new DefaultJobManagerRunnerRegistry(16));
+    }
+
+    private Dispatcher(
+            RpcService rpcService,
+            DispatcherId fencingToken,
+            Collection<JobGraph> recoveredJobs,
+            Collection<JobResult> globallyTerminatedJobs,
+            DispatcherBootstrapFactory dispatcherBootstrapFactory,
+            DispatcherServices dispatcherServices,
+            JobManagerRunnerRegistry jobManagerRunnerRegistry)
+            throws Exception {
+        this(
+                rpcService,
+                fencingToken,
+                recoveredJobs,
+                globallyTerminatedJobs,
+                dispatcherServices.getConfiguration(),
+                dispatcherServices.getHighAvailabilityServices(),
+                dispatcherServices.getResourceManagerGatewayRetriever(),
+                dispatcherServices.getHeartbeatServices(),
+                dispatcherServices.getBlobServer(),
+                dispatcherServices.getFatalErrorHandler(),
+                dispatcherServices.getJobGraphWriter(),
+                dispatcherServices.getJobResultStore(),
+                dispatcherServices.getJobManagerMetricGroup(),
+                dispatcherServices.getMetricQueryServiceAddress(),
+                dispatcherServices.getIoExecutor(),
+                dispatcherServices.getHistoryServerArchivist(),
+                dispatcherServices.getArchivedExecutionGraphStore(),
+                dispatcherServices.getJobManagerRunnerFactory(),
+                dispatcherBootstrapFactory,
+                dispatcherServices.getOperationCaches(),
+                jobManagerRunnerRegistry,
+                new DispatcherResourceCleanerFactory(jobManagerRunnerRegistry, dispatcherServices));
+    }
+
+    @VisibleForTesting
+    protected Dispatcher(
```

Comment:
I still fail to see this simplify things.

- We basically only need to construct dispatcher services once in `TestingDispatcherBuilder#build()`
- We should add new fields `registry` + `cleaner factory` to the dispatcher services to be consistent with the rest

Reply:
The main motivation was to remove this invariant in testing where we have to create an instance of `DispatcherServices` just to pass it into the `TestingDispatcher` and to get all the parameters out again. But considering our discussion in terms of backwards-compatibility - that's probably a perfect example which we shouldn't do to not make life harder for backports... 🤔 
I will refactor it...

Reply:
Registry and cleaner factory cannot be moved into the `DispatcherServices` because of them depending on the Dispatcher's main thread executor. The registry needs to be wrapped into a `OnMainThreadJobManagerRunnerRegistry` which requires the `Dispatcher`'s `MainThreadExecutor` and the `ResourceCleanerFactory` needs the `OnMainThreadJobManagerRunnerRegistry` to ensure that the cleanup is called on the main thread.

I guess letting both, the `JobManagerRunnerRegistry` and the `ResourceCleanerFactory` being instantiated along the `Dispatcher` is still valid considering the tight connection due to the main thread requirement...